api/src/main/java/io/minio/MinioClient.java

```diff
@@ -1209,11 +1214,11 @@ private void updateRegionCache(String bucketName)
            InvalidKeyException, NoResponseException, XmlPullParserException, ErrorResponseException,
            InternalException {
     if (bucketName != null && this.accessKey != null && this.secretKey != null
-        && !BucketRegionCache.INSTANCE.exists(bucketName)) {
+          && !BucketRegionCache.INSTANCE.exists(bucketName)) {
       Map<String,String> queryParamMap = new HashMap<>();
       queryParamMap.put("location", null);
 
-      HttpResponse response = execute(Method.GET, US_EAST_1, bucketName, null,
+      HttpResponse response = execute(Method.GET, US_EAST_1, bucketName, "/",
```

Comment:
With little thought, adding `/` doesn't give a value.  Two cases here.
- path style.  there is no meaning to adding `/`.  its not http specification (I need to verify)
- url style. no need to add in this case.

I think nginix is mis-configured.  Please verify.

Reply:
yes, setting nginx location to `/bucketName` instead of `/bucketName/` will make it everything work as expected. 

However, I verified with other SDKs (minio-go, aws-sdk-go) and they work with both configurations (`/bucketName` and `/bucketName/`), and minio-java should be compatible as well.

Reply:
The problem is perhaps that minio-java is not redirecting correctly, when a `/bucketName` request is made it re-directs to `/bucketName/` but doesn't reAuth the request.

This type of change is not needed if the re-authentication is enabled.

Reply:
> The problem is perhaps that minio-java is not redirecting correctly, when a /bucketName request is made it re-directs to /bucketName/ but doesn't reAuth the request.

minio-java sends the request to `/bucketName` while other SDKs send it to `/bucketName/`. 

When nginx location is set to `/bucketName/` , nginx, irrespective of income request location adds the trailing `/` (if not present), i.e .nginx does the redirection.

Since minio-java calculated signature based on `/bucketname`, it fails due to signature mismatch. Since other SDKs signed based on `/bucketName/`, it works fine.

Reply:
How does aws-sdk-java handle get bucket location query? with `/` or without `/`?

Reply:
let me check and confirm, I had tried with aws-sdk-go .

Reply:
You could try aws-sdk-java with nginix configured with/without `/` on AWS S3 and play.minio.io

Reply:
> aws-sdk-java with ngnix configured with/without / on AWS S3 and play.minio.io

Both upstreams you mentioned are SSL enabled, and sending secure traffic to upstream servers needs detailed configuration and CA certificate (ref: https://www.nginx.com/resources/admin-guide/nginx-tcp-ssl-upstreams/) 

So, I tested with both Nginx configurations `/bucketName` and `/bucketName/` with local upstream Minio server built from latest master, in both the cases `aws-sdk-java` `getBucketLocation` API sends out requests to `bucketName/`, this is Minio server trace

```sh
[REQUEST (objectAPIHandlers).GetBucketLocationHandler-fm] [151223897.018844] [2017-12-02 23:52:50 +0530]
GET /photos/?location
Host: domain.abc
Connection: close
Content-Length: 0
X-Amz-Date: 20171202T182250Z
Amz-Sdk-Retry: 0/0/500
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
Authorization: AWS4-HMAC-SHA256 Credential=minio/20171202//s3/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-retry;content-type;host;user-agent;x-amz-content-sha256;x-amz-date, Signature=a47038e0a9611345c9a745d419951aef434eb54a899c8735289a53a0b5776d44
User-Agent: aws-sdk-java/1.11.240 Linux/4.13.0-16-lowlatency OpenJDK_64-Bit_Server_VM/25.151-b12 java/1.8.0_151
Amz-Sdk-Invocation-Id: 4b598be3-0530-b9b5-222b-7626ca651776
Content-Type: application/octet-stream


[RESPONSE] [151223897.018844] [2017-12-02 23:52:50 +0530]
200 OK
X-Amz-Request-Id: 14FC8D54C99A0977
Server: Minio/DEVELOPMENT.GOGET (linux; amd64)
Accept-Ranges: bytes
Content-Type: application/xml
Vary: Origin
```

Reply:
This means URI ending with `/` comes with a meaning i.e. `createRequest()` should preserve `/`.  This fix needs to be done accordingly.